api/v1alpha3/machinedeployment_webhook.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+func (m *MachineDeployment) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(m).
+		Complete()
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-cluster-x-k8s-io-v1alpha3-machinedeployment,mutating=false,failurePolicy=fail,groups=cluster.x-k8s.io,resources=machinedeployments,versions=v1alpha3,name=validation.machinedeployment.cluster.x-k8s.io
+
+var _ webhook.Validator = &MachineDeployment{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (m *MachineDeployment) ValidateCreate() error {
+	return m.validate()
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (m *MachineDeployment) ValidateUpdate(old runtime.Object) error {
+	return m.validate()
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (m *MachineDeployment) ValidateDelete() error {
+	return nil
+}
+
+func (m *MachineDeployment) validate() error {
+	var allErrs field.ErrorList
+	selector, err := metav1.LabelSelectorAsSelector(&m.Spec.Selector)
+	if err != nil {
+		allErrs = append(
+			allErrs,
+			field.Invalid(field.NewPath("spec", "selector"), m.Spec.Selector, err.Error()),
+		)
+	} else if !selector.Matches(labels.Set(m.Spec.Template.Labels)) {
+		allErrs = append(
+			allErrs,
+			field.Invalid(
+				field.NewPath("spec", "template", "labels"),
+				m.Spec.Template.Labels,
+				"failed validation on label selector, cannot match Machine template labels",
```

Comment:
This prints out as, e.g.:

```
spec.template.labels: Invalid value: "a=b": failed validation on label selector, cannot match Machine template labels
```

If you change the detail message to `fmt.Sprintf("must match spec.selector %q", selector.String()),`, the output looks like this:

```
spec.template.labels: Invalid value: "a=b": must match spec.selector "c=d"
```

WDYT?

Reply:
Yeah, I just tested that out and it is definitely an improvement! It's still verbose but gets the point through.

This is the output I got.

> Error from server (MachineDeployment.cluster.x-k8s.io "test1-md-0" is invalid: spec.template.labels: Invalid value: map[string]string{"nodepool":"nodepool3"}: must match spec.selector "nodepool=nodepool-0"): error when creating "/tmp/md.yaml": admission webhook "validation.machinedeployment.cluster.x-k8s.io" denied the request: MachineDeployment.cluster.x-k8s.io "test1-md-0" is invalid: spec.template.labels: Invalid value: map[string]string{"nodepool":"nodepool3"}: must match spec.selector "nodepool=nodepool-0"